src/bot.rs

```diff
@@ -200,308 +306,234 @@ impl Bot {
-                )
-                .unwrap();
+                self.assign_author(&gh_meta.repo_name, gh_meta.pr_number, github_login)?;
+
+                Some(SystemTime::now())
             }
 
-            for reviewer in pr.requested_reviewers {
-                let reviewer_id_bytes = reviewer.id.to_be_bytes();
-                live_ids.push(reviewer_id_bytes.clone());
-                if db.get_pinned(&reviewer_id_bytes).is_ok() {
-                    let entry = DbEntry::ReviewRequest {
-                        created_at: SystemTime::now(),
-                        reviewer_id: reviewer.id,
-                        repo_name: repo.name.clone(),
-                        reviewer_login: reviewer.login.clone(),
-                        matrix_public_ping_count: 0,
-                    };
-                    db.put(
-                        reviewer_id_bytes,
-                        serde_json::to_string(&entry)
-                            .expect("serialize ReviewRequestEntry")
-                            .as_bytes(),
-                    )
-                    .unwrap();
-                }
+            25..=48 if engineer.riot_id.is_some() => {
+                let riot_id = engineer.riot_id.as_ref().unwrap();
+                // dm author in matrix
+                let message = format!("{author}, {message}", author = riot_id, message = message);
+
+                self.matrix.direct_message(&[&riot_id], &message)?;
+
+                Some(SystemTime::now())
             }
-        }
+
+            49..=72 if engineer.riot_id.is_some() => {
+                let riot_id = engineer.riot_id.as_ref().unwrap();
+                // dm author in matrix
+                let message = format!("{author}, {message}", author = riot_id, message = message);
+                self.matrix.room_message(project_room, &message)?;
+
+                Some(SystemTime::now())
+            }
+
+            _ => last_ping,
+        };
+
+        Ok(last_ping)
     }
 
-    // delete dead entries
-    for (key, _) in db.iterator(IteratorMode::Start) {
-        if !live_ids.iter().any(|id| id == &*key) {
-            db.delete(&key).unwrap();
+    pub fn update_entry(&self, key: &[u8], entry: DbEntry) -> Result<()> {
+        self.db.delete(&key).context(error::Db)?;
+
+        self.db
+            .put(
+                key,
+                serde_json::to_string(&entry)
+                    .context(error::Json)?
+                    .as_bytes(),
+            )
+            .context(error::Db)
+    }
+
+    pub fn act(&mut self) -> Result<()> {
+        for (key, value) in self.db.iterator(IteratorMode::Start) {
+            let entry: DbEntry =
+                serde_json::from_str(&String::from_utf8_lossy(&value)).context(error::Json)?;
+            log::info!("Checking Entry: {:?}", entry);
+
+            match entry {
+                DbEntry::PullRequest(mut data) => {
+                    if data.n_participants >= 2 {
```

Comment:
`n_participants` will currently never be set, right?

Reply:
It is currently set, see lines 462–472.